cql3/statements/select_statement.cc

```diff
@@ -283,33 +283,45 @@ select_statement::make_partition_slice(const query_options& options) const
         std::reverse(bounds.begin(), bounds.end());
         ++_stats.reverse_queries;
     }
+
+    const uint64_t per_partition_limit = get_inner_loop_limit(get_limit(options, _per_partition_limit),
+        options, _per_partition_limit_unset_guard, _selection->is_aggregate(), query::partition_max_rows);
     return query::partition_slice(std::move(bounds),
-        std::move(static_columns), std::move(regular_columns), _opts, nullptr, get_per_partition_limit(options));
+        std::move(static_columns), std::move(regular_columns), _opts, nullptr, per_partition_limit);
 }
 
-uint64_t select_statement::do_get_limit(const query_options& options,
-                                        const std::optional<expr::expression>& limit,
-                                        const expr::unset_bind_variable_guard& limit_unset_guard,
-                                        uint64_t default_limit) const {
-    if (!limit.has_value() || limit_unset_guard.is_unset(options) || _selection->is_aggregate()) {
-        return default_limit;
-    }
-
-    auto val = expr::evaluate(*limit, options);
-    if (val.is_null()) {
-        throw exceptions::invalid_request_exception("Invalid null value of limit");
+select_statement::get_limit_result select_statement::get_limit(
+    const query_options& options, const std::optional<expr::expression>& limit) const
+{
+    if (!limit.has_value()) {
+        return bo::success(query::max_rows);
     }
     try {
+        auto val = expr::evaluate(*limit, options);
+        if (val.is_null()) {
+            return bo::failure(exceptions::invalid_request_exception("Invalid null value of limit"));
+        }
         auto l = val.view().validate_and_deserialize<int32_t>(*int32_type);
         if (l <= 0) {
-            throw exceptions::invalid_request_exception("LIMIT must be strictly positive");
+            return bo::failure(exceptions::invalid_request_exception("LIMIT must be strictly positive"));
         }
-        return l;
+        return bo::success(l);
     } catch (const marshal_exception& e) {
-        throw exceptions::invalid_request_exception("Invalid limit value");
+        return bo::failure(exceptions::invalid_request_exception("Invalid limit value"));
+    } catch (const exceptions::invalid_request_exception& e) {
+        return bo::failure(e);
     }
 }
 
+uint64_t select_statement::get_inner_loop_limit(const select_statement::get_limit_result& limit, const query_options& options,
+    const expr::unset_bind_variable_guard& unset_guard, bool is_aggregate, uint64_t default_value)
+{
+    if (!limit.has_value() || unset_guard.is_unset(options) || is_aggregate) {
+        return default_value;
+    }
+    return limit.value();
```

Comment:
I don't think you need to pass options/unset_guard/default_value here. get_limit() substitutes max_rows if not specified, which is equal to default_value in all cases.

In one case you pass query::partition_max_rows rather than query::max_rows. The question is whether they is accidentally equal or conceptually equal. I think they are conceptually equal (basically in both cases signifying an infinite amount). Maybe the reason we have two is 45215746fe6c013047b69aa651eb1a203a8c4185.

So you can simplify it to check only is_aggregate.